### PR TITLE
Block Library: Add `EntityProvider`, controlled `InnerBlocks` and basic template blocks.

### PR DESCRIPTION
**Pull request by epiqueras**
cc @youknowriad 

## Description

This PR puts together the best ideas in #17020 and #17118 to implement Full-Site editing blocks.

Here is the prior art of this refined approach: https://github.com/WordPress/gutenberg/pull/17118#issuecomment-523621576.

The Site Title Block was intentionally left out to reduce the scope of this PR, but the Post Content Block was kept so that we have a block to use the new "semi-controlled" `InnerBlocks` with.

## How has this been tested?

The blocks were tested manually to work as expected.

## Types of Changes

*New Feature:* Add new template blocks: Post, Post Title, and Post Content.

## Checklist:

- [ ] My code is tested.
- [x] My code follows the WordPress code style. 
- [x] My code follows the accessibility standards. 
- [x] My code has proper inline documentation. 
- [x] I've included developer documentation if appropriate.

### CODE REVIEW

File: packages/core-data/src/entity-provider.js [hunk 1]
```diff
@@ -0,0 +1,71 @@
...
+ * Internal dependencies
+ */
+import { defaultEntities } from './entities';
+
+const entities = {
+	...defaultEntities.reduce( ( acc, entity ) => {
+		acc[ entity.name ] = { kind: entity.kind, context: createContext() };
+		return acc;
+	}, {} ),
+	post: { kind: 'postType', context: createContext() },
```
Comment (youknowriad) [inline packages/core-data/src/entity-provider.js]: If I remember properly, there's no way to ensure the name is unique across kinds. so we might be better using a two level objects [kind][name] or just an array of entities.

Also, there might be a better way to automatically create contexts for these dynamic entities (like post).

**Review by youknowriad — COMMENTED** [review]
I agree that API-wise, these seems like a good approach. I'm not sure I understand the changes in InnerBlocks entirely yet and as I said, I'm still convinced that we might be better served with separate block lists (without intersections) for separate entities. For example here  if I serialize the blocks for a given entity, I'm pretty certain the serialization will include the blocks from other entities in the produces HTML (post-content block).

**Comment by epiqueras** [comment]
>I'm not sure I understand the changes in InnerBlocks entirely yet

It's just allowing the parent to sync with it through value and on change props.

>I'm pretty certain the serialization will include the blocks from other entities in the produces HTML (post-content block).

No, that's explicitly controlled by the post content block not having a save component.

File: packages/core-data/src/entity-provider.js (hunk 1 above)
Comment (epiqueras) [reply packages/core-data/src/entity-provider.js]: >be better using a two level objects [kind][name]

Yes

>Also, there might be a better way to automatically create contexts for these dynamic entities (like post).

We could add this to the store and load them asynchronously, but it seems unnecessarily complex.

File: packages/core-data/src/entity-provider.js [hunk 2]
```diff
@@ -0,0 +1,71 @@
...
+	const value = useSelect(
+		( select ) => {
+			const entity = select( 'core' ).getEditedEntityRecord( kind, type, id );
+			return entity && entity[ prop ];
+		},
+		[ kind, type, id, prop ]
+	);
+
+	const { editEntityRecord } = useDispatch( 'core' );
+	const setValue = useCallback(
```
[the remark this answers is not in the document]
Comment (epiqueras) [reply packages/core-data/src/entity-provider.js]: This is a library hook, so the cost will be determined by the user, unknown to us.

Users don't expect the function returned to change (like with `useState`), and they might pass it directly to pure components and break their optimization.

The cost of `useCallback` is minimal even if paid everywhere the hook is used. A single pure component de-optimization probably offsets it.

```
useCallbackCost * nTimesUsed < potentiallyLargePureComponentDeOptimizationCost * probabilityThisHappens
```

File: packages/core-data/src/entity-provider.js (hunk 1 above)
Comment (epiqueras) [reply packages/core-data/src/entity-provider.js]: Or.....

https://github.com/WordPress/gutenberg/pull/17153/commits/217bcad354421952a0508342d46c2ba1fc5fcdd1

:rocket::rocket::rocket::rocket::rocket::rocket::rocket::rocket::rocket::laughing:

**Pull request closed without merging**